pkg/csi_driver/reconciler.go

```diff
@@ -165,14 +174,349 @@ func (recon *multishareReconciler) reconcileWorker() {
+				continue
+			}
+			if shareInfo.Status.ShareStatus == v1alpha1.DELETED {
+				klog.Infof("shareInfo status DELETED, no need to send any share request for %s", shareInfo.Name)
+				continue
+			}
+
+			if !shareExist(shareInfo, instanceShares) {
+				newInstanceInfo, err := recon.maybeMarkShareInfoStatusDeleted(shareInfo, instanceInfos[shareInfo.Status.InstanceHandle])
+				if err != nil {
+					klog.Errorf("failed to mark shareInfo %s deleted : %s", shareInfo.Name, err.Error())
+					continue
+				}
+				instanceInfos[shareInfo.Status.InstanceHandle] = newInstanceInfo
+				continue
+			}
+		}
+
+		if shareInfo.Status == nil || shareInfo.Status.InstanceHandle == "" {
+			klog.Infof("share %s not assigned to any instance yet", shareInfo.Name)
+			continue
+		}
+
+		if !needDelete && shareInfo.Spec.CapacityBytes == shareInfo.Status.CapacityBytes {
+			klog.V(6).Infof("no need to send any share request for %s", shareInfo.Name)
+			continue
+		}
+
+		assignedInstanceInfo, ok := instanceInfos[shareInfo.Status.InstanceHandle]
+		if !ok {
+			klog.Errorf("share %s is assigned to %s but instanceInfo does not exist", shareInfo.Name, shareInfo.Status.InstanceHandle)
+			continue
+		}
+		if assignedInstanceInfo.Status == nil || assignedInstanceInfo.Status.InstanceStatus != v1alpha1.READY || assignedInstanceInfo.Status.CapacityBytes < assignedInstanceInfo.Spec.CapacityBytes {
+			klog.Infof("InstanceStatus for %s is not ready", assignedInstanceInfo.Name)
+			if assignedInstanceInfo.Status != nil && assignedInstanceInfo.Status.Error != "" {
+				recon.updateShareInfoErr(shareInfo, fmt.Errorf(assignedInstanceInfo.Status.Error))
+			}
+			continue
+		}
+
+		project, instanceRegion, name, err := util.ParseInstanceURI(shareInfo.Status.InstanceHandle)
+		if err != nil {
+			klog.Errorf("failed to parse instanceURI %q: %s", shareInfo.Status.InstanceHandle, err.Error())
+			continue
+		}
+		share := &file.Share{
+			Name: shareInfo.Spec.ShareName,
+			Parent: &file.MultishareInstance{
+				Project:  project,
+				Location: instanceRegion,
+				Name:     name,
+			},
+			CapacityBytes:  shareInfo.Spec.CapacityBytes,
+			MountPointName: shareInfo.Spec.ShareName,
+			Labels:         shareInfo.Labels,
+		}
+
+		shareURI, err := file.GenerateShareURI(share)
+		if err != nil {
+			klog.Errorf("error generating shareURI for %s: %s", shareInfo.Name, err.Error())
+		}
+		op, err := runningOpMaybeErrForTarget(shareURI, ops)
+		if err != nil {
+			recon.updateShareInfoErr(shareInfo, err)
+		}
+
+		if op == nil {
+			klog.Infof("no running Op found for %s", shareURI)
+			if needDelete {
+				// delete
+				klog.Infof("Starting share Delete operation for %s", shareURI)
+				_, err = recon.cloud.File.StartDeleteShareOp(context.TODO(), share)
+			} else if shareInfo.Status.ShareStatus != v1alpha1.READY {
+				// create
+				klog.Infof("Starting share Create operation for %s", shareURI)
+				_, err = recon.cloud.File.StartCreateShareOp(context.TODO(), share)
```

Comment:
if we are only looking at status != READY, will we attempt create share on every round of reconciliation?

Reply:
in L282 we check `op==nil`, so only when there's no existing OP and status is not `READY` we'll send out create share requests.